numpy/core/tests/test_multiarray.py

```diff
@@ -1357,6 +1357,34 @@ def test_multiindex_titles(self):
         assert_raises(ValueError, lambda : a[['b','b']])  # field exists, but repeated
         a[['b','c']]  # no exception
 
+    def test_structured_cast_promotion_fieldorder(self):
+        # gh-15494
+        # dtypes with different field names are not promotable
+        A = ("a", "<i8")
+        B = ("b", "<i8")
+        ab = np.array([(1, 2)], dtype=[A, B])
+        ba = np.array([(1, 2)], dtype=[B, A])
+        assert_raises(TypeError, np.concatenate, [ab, ba])
+        assert_raises(TypeError, np.result_type, [ab.dtype, ba.dtype])
+        assert_raises(TypeError, np.promote_types, [ab.dtype, ba.dtype])
+
+        # dtypes with same field names/order but different memory offsets
+        # are not promotable either.
+        assert_raises(TypeError, np.promote_types,
+                                 [ab.dtype, ba[['a', 'b']].dtype])
```

Comment:
I imagine this is going to be a bit annoying for when I try to do something like `np.concatenate((arr1[['a', 'b']], arr2[['a', 'b']]))`.

A reasonable alternative might be for `promote_types` to eliminate all padding if it doesn't match.

But maybe we can deal with that in a follow-up.

Reply:
That seems reasonable. I think a counterargument might be that we want to avoid adding special behaviors/cases the user has to know about, and should instead to force the user to be specific, eg by doing
```python
np.concatenate((repack_fields(arr1[['a', 'b']]), repack_fields(arr2[['a', 'b']])))
```

Better yet, I think users should use something like a ducktyped-ArrayCollection instead of a structured array where padding is important.

In any case, let's disallow it in this PR.

Reply:
Rethinking, I actually would argue that promote types *should* eliminate all paddings. Since it normally returns a *new* dtype that can represent the same information as both input dtypes (safely cast both of them to this one).
There is some complexity about changing the padding when both input dtypes are identical (i.e. we currently ensure native byte order after promotion, but is a bit dubious).

In the future, I am thinking about adding a `to_canonical()` method, or so, which I suppose for structured dtypes could strip padding.

I have to think about what this actually changes API wise, I expect I think this fixes cases which are probably bugs only?

Reply:
The `promote_types` docstring does say "Returns the data type with the smallest size and smallest scalar kind ", which suggests that if you do `promote_types(a, a)` where a contains padding, it would remove the padding this would lead to the "smallest size".

So if we are dogmatically following the docstring, it does make sense to remove padding.

Reply:
I fully agree, but I would not give much heed to the docstring, it was obviously written with numerical types in mind :).